File: ios/Classes/RewardedAd.m
#import "FlutterAdsApplovinMaxPlugin.h"
#import "RewardedAd.h"
#import <objc/runtime.h>

@implementation RewardedAd {
  MARewardedAd* rewarded;
  int retryAttempt;
}

- (void)initialize:(NSString*)unitId {
  rewarded = [MARewardedAd sharedWithAdUnitIdentifier: unitId];
  rewarded.delegate = self;

  // Load the first ad
  [rewarded loadAd];
}

- (BOOL)isLoaded {
  return [rewarded isReady];
}

- (void)show {
  if ( [rewarded isReady] )
  {
      [rewarded showAd];
  }
}

#pragma mark - MAAdDelegate Protocol

- (void)didLoadAd:(MAAd *)ad
{
    // Rewarded ad is ready to be shown. '[self.rewardedAd isReady]' will now return 'YES'

    // Reset retry attempt
    retryAttempt = 0;
    [globalChannel invokeMethod:@"onRewardedAdLoaded" arguments:nil];
}

- (void)didFailToLoadAdForAdUnitIdentifier:(NSString *)adUnitIdentifier withError:(MAError *)error
{
    // Rewarded ad failed to load
    // We recommend retrying with exponentially higher delays up to a maximum delay (in this case 64 seconds)

    retryAttempt++;
    NSInteger delaySec = pow(2, MIN(6, retryAttempt));

[globalChannel invokeMethod:@"onRewardedAdLoadFailed" arguments:nil];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, delaySec * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        [self->rewarded loadAd];
    });
}

- (void)didDisplayAd:(MAAd *)ad {
[globalChannel invokeMethod:@"onRewardedAdDisplayed" arguments:nil];
}

- (void)didClickAd:(MAAd *)ad {
[globalChannel invokeMethod:@"onRewardedAdClicked" arguments:nil];
}

- (void)didHideAd:(MAAd *)ad
{
[globalChannel invokeMethod:@"onRewardedAdHidden" arguments:nil];
    // Rewarded ad is hidden. Pre-load the next ad
    [rewarded loadAd];
}

- (void)didFailToDisplayAd:(MAAd *)ad withError:(MAError *)error
{
    // Rewarded ad failed to display. We recommend loading the next ad

    [globalChannel invokeMethod:@"onRewardedAdDisplayFailed" arguments:nil];
    [rewarded loadAd];
}

#pragma mark - MARewardedAdDelegate Protocol

- (void)didStartRewardedVideoForAd:(MAAd *)ad {
[globalChannel invokeMethod:@"onRewardedAdStarted" arguments:nil];
}

- (void)didCompleteRewardedVideoForAd:(MAAd *)ad {
[globalChannel invokeMethod:@"onRewardedAdCompleted" arguments:nil];
}

- (void)didRewardUserForAd:(MAAd *)ad withReward:(MAReward *)reward
{
    // Rewarded ad was displayed and user should receive the reward
    retryAttempt = 0;
    [globalChannel invokeMethod:@"onUserRewarded" arguments:nil];
}

@end

File: android/src/main/java/net/smartphone/games/flutter_ads_applovin_max/RewardedAd.java
package net.smartphone.games.flutter_ads_applovin_max;

import com.applovin.mediation.MaxAd;
import com.applovin.mediation.MaxAdRevenueListener;
import com.applovin.mediation.MaxError;
import com.applovin.mediation.MaxReward;
import com.applovin.mediation.MaxRewardedAdListener;
import com.applovin.mediation.ads.MaxRewardedAd;

import io.flutter.Log;

public class RewardedAd implements MaxRewardedAdListener, MaxAdRevenueListener {
    private MaxRewardedAd reward;
    private int retryAttempt;

    public void initialize(String unitId) {
        reward = MaxRewardedAd.getInstance(unitId, FlutterAdsApplovinMaxPlugin.getInstance().activity );
        reward.setListener( this );
        reward.loadAd();
    }

    public void show() {
        try {
            if (reward != null && reward.isReady() && FlutterAdsApplovinMaxPlugin.getInstance().activity != null)
                reward.showAd();
        } catch (Exception e) {
            Log.e("AppLovin", e.toString());
        }
    }

    public boolean isLoaded() {
        return reward.isReady();
    }

    @Override
    public void onAdLoaded(MaxAd ad) {
        retryAttempt = 0;
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdLoaded");

    }

    @Override
    public void onAdDisplayed(MaxAd ad) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdDisplayed");

    }

    @Override
    public void onAdHidden(MaxAd ad) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdHidden");
        reward.loadAd();
    }

    @Override
    public void onAdClicked(MaxAd ad) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdClicked");
    }

    @Override
    public void onAdLoadFailed(String adUnitId, MaxError error) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdLoadFailed");
    }

    @Override
    public void onAdDisplayFailed(MaxAd ad, MaxError error) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdDisplayFailed");
    }

    @Override
    public void onRewardedVideoStarted(MaxAd ad) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdStarted");

    }

    @Override
    public void onRewardedVideoCompleted(MaxAd ad) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdCompleted");

    }

    @Override
    public void onAdRevenuePaid(MaxAd ad) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onRewardedAdRevenuePaid");
    }

    @Override
    public void onUserRewarded(MaxAd ad, MaxReward reward) {
        FlutterAdsApplovinMaxPlugin.getInstance().sendToFlutter("onUserRewarded");
    }
}